### PR TITLE
Fix #45213: Syncing jobs would return error when podController exception

### LINKED ISSUE #45213: BUG: Memory-limited jobs with a resource quota that initially don't have enough resources stay in queue almost forever
**What keywords did you search in Kubernetes issues before filing this one?**
Job, resourcequota, scheduler

- **Kubernetes version** (use `kubectl version`):

    Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1", GitCommit:"b0b7a323cc5a4a2019b2e9520c21c7830b7f708e", GitTreeState:"clean", BuildDate:"2017-04-03T20:44:38Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
    Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"a55267932d501b9fbd6d73e5ded47d79b5763ce5", GitTreeState:"clean", BuildDate:"2017-04-14T13:36:25Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}

- **Environment**: CentOS bare metal (https://kubernetes.io/docs/getting-started-guides/centos/centos_manual_config/)
- **Cloud provider or hardware configuration**: -
- **OS** (e.g. from /etc/os-release): CentOS 7
- **Kernel** (e.g. `uname -a`): Linux <server name> 3.10.0-514.2.2.el7.x86_64 #1 SMP Tue Dec 6 23:06:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**: -
- **Others**: -

**What happened**:
Having created a resourcequota (e.g. 400M) and creating enough memory-limited jobs to go over that quota (e.g. 4 x 200M limit jobs), at first, two jobs are created and executed as expected, and two jobs fail at first due to memory quota:

    Events:
      FirstSeen	LastSeen	Count	From		SubObjectPath	Type		Reason		Message
      ---------	--------	-----	----		-------------	--------	------		-------
      4s		4s		1	job-controller			Warning		FailedCreate	Error creating: pods "memhog-y152c55u-" is forbidden: exceeded quota: memoryquota, requested: memory=200Mi, used: memory=400Mi, limited: memory=400Mi

After the first two jobs have been completed, instead of starting the remaining jobs as soon as possible, the remaining two jobs stay unstarted in this state for many, many hours until they finally start and complete.

First job: Start Time:	Fri, 28 Apr 2017 15:00:08 +0300
One of the remaining 2 jobs: Start Time:	Sat, 29 Apr 2017 07:13:34 +0300

**What you expected to happen**:

I'm expecting the two waiting jobs to start executing as soon as the resources they need are available, e.g. as soon as the first jobs are completed.

**How to reproduce it** (as minimally and precisely as possible):

Bring up a kubernetes cluster, then execute these commands:

    $ kubectl create quota memoryquota --hard=memory=400Mi
    $ echo kubectl run memhog-'$(cat /dev/urandom | tr -dc 'a-z0-9' | fold -w 8 | head -n 1)' --replicas=1 --restart=OnFailure --image-pull-policy=IfNotPresent --limits=memory='$1'Mi,cpu=100m --record --image=derekwaynecarr/memhog --command -- memhog -r100 '$2'm > run-memhog.sh
    $ chmod +x run-memhog.sh
    $ for i in {1..4}; do ./run-memhog.sh 200 100; done
    job "memhog-0ju2eiuy" created
    job "memhog-t8qyt5sj" created
    job "memhog-ljaqz4ek" created
    job "memhog-eb8q8kr3" created

Wait for around 30 seconds and the first jobs should be complete:
    
    $ kubectl get jobs
    NAME              DESIRED   SUCCESSFUL   AGE 
    memhog-0ju2eiuy   1         1            43s
    memhog-eb8q8kr3   1         0            42s
    memhog-ljaqz4ek   1         0            43s
    memhog-t8qyt5sj   1         1            43s
    $ kubectl describe jobs
    ....

However, the remaining jobs don't start at all as shown by:

    $ kubectl describe job memhog-ljaqz4ek
    Name:   memhog-ljaqz4ek
    Namespace:  test
    Selector: controller-uid=dd6fb770-2f00-11e7-8766-52c84ff43b2b
    Labels:   run=memhog-ljaqz4ek
    Annotations:  kubernetes.io/change-cause=kubectl run memhog-ljaqz4ek memhog -r100 100m --replicas=1 --restart=OnFailure --image-pull-policy=IfNotPresent --limits=memory=200Mi,cpu=100m --record=true --image=derekway...
    Parallelism:  1
    Completions:  1
    Pods Statuses:  0 Running / 0 Succeeded / 0 Failed
    Pod Template:
      Labels: controller-uid=dd6fb770-2f00-11e7-8766-52c84ff43b2b
        job-name=memhog-ljaqz4ek
        run=memhog-ljaqz4ek
      Containers:
       memhog-ljaqz4ek:
        Image:  derekwaynecarr/memhog
        Port: 
        Command:
          memhog
          -r100
          100m
        Limits:
          cpu:    100m
          memory:   200Mi
        Environment:  <none>
        Mounts:   <none>
      Volumes:    <none>
    Events:
      FirstSeen LastSeen  Count From    SubObjectPath Type    Reason    Message
      --------- --------  ----- ----    ------------- --------  ------    -------
      3m    3m    1 job-controller      Warning   FailedCreate  Error creating: pods "memhog-ljaqz4ek-" is forbidden: exceeded quota: memoryquota, requested: memory=200Mi, used: memory=400Mi, limited: memory=400Mi    

Furthermore, pods related to the remaining jobs aren't even created and only the completed pods are shown:

    $ kubectl get pods -a
    NAME                    READY     STATUS      RESTARTS   AGE
    memhog-0ju2eiuy-bz0zm   0/1       Completed   0          7m
    memhog-t8qyt5sj-q9msw   0/1       Completed   0          7m    

**Anything else we need to know**:

I've also created a stackoverflow question about this (http://stackoverflow.com/questions/43654683/resourcequota-multiple-memory-limited-jobs-restarting-pending-jobs-takes-for), and asked about this in the kubernetes slack, where I was urged to create a bug report.

### PR DESCRIPTION
**Pull request by tacy**
**What this PR does / why we need it**:
Jobcontroller:  Syncing jobs would return error when podController exception
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes #45213 
**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/controller/job/jobcontroller.go [hunk 1]
```diff
@@ -437,24 +439,36 @@ func (jm *JobController) syncJob(key string) error {
...
 		// kill remaining active pods
 		wait := sync.WaitGroup{}
+		errCh := make(chan error, int(active))
 		wait.Add(int(active))
 		for i := int32(0); i < active; i++ {
 			go func(ix int32) {
 				defer wait.Done()
 				if err := jm.podControl.DeletePod(job.Namespace, activePods[ix].Name, &job); err != nil {
-					defer utilruntime.HandleError(err)
+					defer wait.Done()
```
Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: You don't need to do it 2nd time, it's already done 2 lines above.

File: pkg/controller/job/jobcontroller.go [hunk 2]
```diff
@@ -437,24 +439,36 @@ func (jm *JobController) syncJob(key string) error {
...
+					glog.V(2).Infof("Failed to delete %v, job %q/%q deadline exceeded", activePods[ix].Name, job.Namespace, job.Name)
```
Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: Please leave the `utilruntime.HandleError`, it does glog, but can additionally feed other error handlers with information.

File: pkg/controller/job/jobcontroller.go [hunk 3]
```diff
@@ -536,18 +550,20 @@ func getStatus(pods []*v1.Pod) (succeeded, failed int32) {
 // manageJob is the core method responsible for managing the number of running
 // pods according to what is specified in the job.Spec.
 // Does NOT modify <activePods>.
-func (jm *JobController) manageJob(activePods []*v1.Pod, succeeded int32, job *batch.Job) int32 {
+func (jm *JobController) manageJob(activePods []*v1.Pod, succeeded int32, job *batch.Job) (active int32, err error) {
```
Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: Drop named return parameters, you're not using it at all.

File: pkg/controller/job/jobcontroller.go [hunk 4]
```diff
@@ -562,12 +578,13 @@ func (jm *JobController) manageJob(activePods []*v1.Pod, succeeded int32, job *b
 			go func(ix int32) {
 				defer wait.Done()
 				if err := jm.podControl.DeletePod(job.Namespace, activePods[ix].Name, job); err != nil {
-					defer utilruntime.HandleError(err)
 					// Decrement the expected number of deletes because the informer won't observe this deletion
+					glog.V(2).Infof("Failed to delete %v, decrementing expectations for job %q/%q", activePods[ix].Name, job.Namespace, job.Name)
```
Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: Same, leave `utilruntime.HandleError`.

File: pkg/controller/job/jobcontroller.go [hunk 5]
```diff
@@ -607,19 +625,29 @@ func (jm *JobController) manageJob(activePods []*v1.Pod, succeeded int32, job *b
...
 				}
 			}()
 		}
 		wait.Wait()
 	}
 
-	return active
+	select {
+	case err := <-errCh:
+		// all errors have been reported before and they're likely to be the same, so we'll only return the first one we hit.
```
Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: That is not a good assumption. Not every error will be the same.

**Review by soltysh — CHANGES_REQUESTED** [review]

**Comment by soltysh** [comment]
@tacy reading through the PR I'm not quite sure how this will help you with resolving your problem from #45213. You reported you've had problems with creating pods, but you've only fixed pod removal part, was the removal blocking you? I'd expect [this place](https://github.com/kubernetes/kubernetes/blob/7b0dee89f29c4c9e645f3d66c98aeb2250783e0a/pkg/controller/job/jobcontroller.go#L609) to handle creation failure and do not increase the active count until one of the currently running pods finishes. Lemme try that example of yours and see what I'm getting...

**Comment by soltysh** [comment]
@derekwaynecarr fyi

**Comment by soltysh** [comment]
Nvmd my last comment, I think I got blind when reading your PR and missed the part wrt to pod creation, sorry 😊

File: pkg/controller/job/jobcontroller.go (hunk 5 above)
Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: I'd drop this comment. Generally what we need is to inform the controller that there was an error and it should re-try this job once more next time.

File: pkg/controller/job/jobcontroller.go [hunk 6]
```diff
@@ -607,19 +625,29 @@ func (jm *JobController) manageJob(activePods []*v1.Pod, succeeded int32, job *b
 			go func() {
 				defer wait.Done()
 				if err := jm.podControl.CreatePodsWithControllerRef(job.Namespace, &job.Spec.Template, job, newControllerRef(job)); err != nil {
-					defer utilruntime.HandleError(err)
```
Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: Same, leave `utilruntime.HandleError`.

**Comment by soltysh** [comment]
@tacy you need to make sure to sign the CLA, as well

**Comment by tacy** [comment]
Thanks for the review @soltysh, comments addressed,  please take a look.

I have signed the CLA.

**Review by soltysh — APPROVED** [review]

**Comment by soltysh** [comment]
@tacy - awesome, thanks! Can you please squash your changes into a single commit?
@k8s-bot ok to test

**Comment by soltysh** [comment]
@tacy wrt to CLA can you comment in https://github.com/kubernetes/kubernetes/issues/27796 and ping `@sarahnovotny` and `@foxish`

**Comment by soltysh** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by tacy** [comment]
@soltysh  - Squash done

**Comment by soltysh** [comment]
/lgtm
/approve

@tacy all is left is the CLA.

**Comment by tacy** [comment]
I have signed, and will call helpdesk<helpdesk@rt.linuxfoundation.org> to find why

thanks @soltysh

**Comment by soltysh** [comment]
@tacy you're fast, right after your comment the bot added the cla approved label, yay! Thanks once again, it should merge shortly :)

**Comment by soltysh** [comment]
@tacy mind cherry-picking this one onto 1.6 branch?

**Comment by tacy** [comment]
Cherry-pink done.

/cc @soltysh

**Pull request merged**